File: backend/src/main/java/django/Signer.java
package django;

public class Signer {
    protected final String SEPARATOR = ":";

    protected final String key;
    protected final String salt;

    public Signer(String key, String salt) {
        this.key = key;
        this.salt = salt;
    }

    public String signature(String value) {
        return base64Hmac(salt + "signer", value.getBytes(), key);
    }

    public String unsign(String signedValue) {
        int sepIndex = signedValue.lastIndexOf(SEPARATOR);
        if (sepIndex < 0) {
            throw new BadSignature("No \"" + SEPARATOR + "\" found in value");
        }
        String value = signedValue.substring(0, sepIndex);
        String sig = signedValue.substring(sepIndex + SEPARATOR.length());
        if (Crypto.constantTimeCompare(sig, signature(value))) {
            return value;
        }
        throw new BadSignature("Signature \"" + sig + "\" does not match");
    }

    private static String base64Hmac(String salt, byte[] value, String key) {
        return Signing.BASE64.encode(Crypto.saltedHmac(salt, value, key));
    }
}

File: backend-ts/src/auth/django/signer.ts
import { createHash, createHmac, timingSafeEqual } from "crypto"
import base64url from "base64url"

class BadSignature extends Error {}

export class Signer {
  protected static SEPARATOR = ":"

  constructor(protected key: string, protected salt: string) {}

  private signature(value: string): string {
    return Signer.base64Hmac(this.salt + "signer", Buffer.from(value), this.key)
  }

  public sign(value: string): string {
    return value + Signer.SEPARATOR + this.signature(value)
  }

  public unsign(signedValue: string): string {
    const sepIndex = signedValue.lastIndexOf(Signer.SEPARATOR)
    if (sepIndex < 0) {
      throw new BadSignature("No \"" + Signer.SEPARATOR + "\" found in value")
    }
    const value = signedValue.substring(0, sepIndex)
    const sig = signedValue.substring(sepIndex + Signer.SEPARATOR.length)
    if (timingSafeEqual(Buffer.from(sig), Buffer.from(this.signature(value)))) {
      return value
    }
    throw new BadSignature("Signature \"" + sig + "\" does not match")
  }

  private static base64Hmac(salt: string, value: Buffer, key: string): string {
    const hash = createHash('sha1').update(salt + key).digest()
    const hmac = createHmac('sha1', hash).update(value).digest()
    return base64url.encode(hmac)
  }
}
